src/librustc/hir/mod.rs

```diff
@@ -2460,6 +2484,19 @@ pub struct Item {
     pub span: Span,
 }
 
+impl Item {
+    /// If this `Item` is a function, return the span of the header, otherwise
+    /// return the span of the whole item.
+    pub fn fn_header_span(&self) -> Span {
```

Comment:
IMO most items could be partitioned into "header" and "body" (either literally `hir::Body` or something else like the RHS of a `type` alias or the fields of a `struct` definition).

I've wanted something like that, alongside `def_span`, for a long time now, so I'd be glad if we could get that - most of the time, using `def_span` (or e.g. an item's span) can result in larger snippet output, of which the header is the main point of interest anyway.

cc @rust-lang/compiler

Reply:
There's a _separate_ `def_span` method which takes a span and returns a smaller span until the first `{`, which is used in some places for this, but I would prefer these new methods to be used for that.

Reply:
Such a method shouldn't be called `def_span` IMO (because of the collision with `tcx.def_span(def_id)`), I'm surprised one exists already.

Reply:
I didn't know about `def_span` either, and had a similar hack in clippy, but was worried that someday const generics with blocks might break them.

<Rant>Rust is getting so complex these days...</Rant>

Reply:
It also predates `Ident`s having spans which I tried to add before and had been shot down due to size concerns. Now all `Ident`s have a `Span`, making `tcx.def_span` unnecessary for most cases.

Reply:
> making `tcx.def_span` unnecessary for most cases.

But they are completely different things, `tcx.def_span(def_id)` is the only general way to get the `Span` of `def_id`, any other methods being more limited, one way or another.

I'd prefer we remove other `Span` functionality, in favor of queries like `tcx.def_span`, when it's intended for (a subset of) entities that have `DefId`s.

Reply:
Ah, pretty sure the former is *not* `tcx.def_span` (the latter is).